new-tutorial/L4_message_passing.py

```diff
@@ -0,0 +1,386 @@
+        Output feature size.
+    """
+    def __init__(self, in_feat, out_feat):
+        super().__init__()
+        # A linear submodule for projecting the input and neighbor feature to the output.
+        self.linear = nn.Linear(in_feat * 2, out_feat)
+    
+    def forward(self, g, h):
+        """Forward computation
+        
+        Parameters
+        ----------
+        g : Graph
+            The input graph.
+        h : Tensor
+            The input node feature.
+        """
+        with g.local_scope():
+            g.ndata['h'] = h
+            # update_all is a message passing API.
+            g.update_all(message_func=fn.copy_u('h', 'm'), reduce_func=fn.mean('m', 'h_N'))
+            h_N = g.ndata['h_N']
+            h_total = torch.cat([h, h_N], dim=1)
+            return self.linear(h_total)
+
+
+######################################################################
+# **First step**: Check whether the input feature is a single tensor or a
+# pair of tensors:
+# 
+# .. code:: python
+# 
+#    if isinstance(h, tuple):
+#        h_src, h_dst = h
+#    else:
+#        h_src = h_dst = h
+# 
+# **Second step**: Replace node features ``h`` with ``h_src`` or
+# ``h_dst``, and assign the node features to ``srcdata`` or ``dstdata``,
+# instead of ``ndata``.
+# 
+# Whether to assign to ``srcdata`` or ``dstdata`` depend on whether the
+# said feature acts as the features on source nodes or destination nodes
+# of the edges in the message functions (in ``update_all`` or
+# ``apply_edges``).
+# 
+# *Example 1*: For the following ``update_all`` statement:
+# 
+# .. code:: python
+# 
+#    g.ndata['h'] = h
+#    g.update_all(message_func=fn.copy_u('h', 'm'), reduce_func=fn.mean('m', 'h_N'))
+# 
+# The node feature ``h`` acts as source node feature because ``'h'``
+# appeared as source node feature. So you will need to replace ``h`` with
+# source feature ``h_src`` and assign to ``srcdata`` for the version that
+# works with both cases:
+# 
+# .. code:: python
+# 
+#    g.srcdata['h'] = h_src
+#    g.update_all(message_func=fn.copy_u('h', 'm'), reduce_func=fn.mean('m', 'h_N'))
+# 
+# *Example 2*: For the following ``apply_edges`` statement:
+# 
+# .. code:: python
+# 
+#    g.ndata['h'] = h
+#    g.apply_edges(fn.u_dot_v('h', 'h', 'score'))
+# 
+# The node feature ``h`` acts as both source node feature and destination
+# node feature. So you will assign ``h_src`` to ``srcdata`` and ``h_dst``
+# to ``dstdata``:
+# 
+# .. code:: python
+# 
+#    g.srcdata['h'] = h_src
+#    g.dstdata['h'] = h_dst
+#    # The first 'h' corresponds to source feature (u) while the second 'h' corresponds to destination feature (v).
+#    g.apply_edges(fn.u_dot_v('h', 'h', 'score'))
+# 
+# .. note::
+# 
+#    For homogeneous graphs (i.e. graphs with only one node type
+#    and one edge type), ``srcdata`` and ``dstdata`` are aliases of
+#    ``ndata``. So you can safely replace ``ndata`` with ``srcdata`` and
+#    ``dstdata`` even for full-graph training.
```

Comment:
We should move this note block to earlier.

Reply:
But the fact that you can safely replace `ndata` with `srcdata`/`dstdata` is only necessary when you actually write a GNN module that works for both cases?  Not sure where you want to move it.